### PR TITLE
feat: search for public keys

### PR DESCRIPTION
**Pull request by ajpauwels**
Part of adding PKI support.

- Adds config support for specifying public and private key folders
- Adds logic for filtering directory contents to create strong rules on naming schemes for keys
- Currently simply outputs valid paths to public keys

### CODE REVIEW

File: src/file_io.rs
Comment (howhelan) [inline src/file_io.rs:31]: I'm assuming you can't overload w Rust?

File: src/file_io.rs
Comment (howhelan) [inline src/file_io.rs:60]: Is `<<` a shift?  Why?

Comment (howhelan) [reply src/file_io.rs:60]: Is bit manipulation really necessary here in general?  Seems like all this could be done more in a more readable way with boolean operations.

File: src/file_io.rs
Comment (howhelan) [reply src/file_io.rs:14]: ok I think I see - it makes it possible to create a default implementation of `FsFilterer` which uses the std fs library?

Comment (ajpauwels) [reply src/file_io.rs]: I just added a lot more. The purpose is to interface out access to the underlying fs module. Initial motivation was that I can't mock out the fs calls because it's all concrete types, so I made traits with the methods I needed and then added pass-through implementations which are used by default in running code. In tests though, I can make mock implementations of those traits and inject them in instead of the fs module.

I really like this pattern in general, I use it on all my data types. More than just tests, when everything is interfaced, it makes it really easy to then remove or rewrite parts of it without affecting other code.

It is more verbose though. Let me know what you think.

File: src/file_io.rs
Comment (ajpauwels) [reply src/file_io.rs]: No, and I really hate how i ended up doing this, just don't know any other simple way to inject the dependency.

File: src/file_io.rs
Comment (ajpauwels) [reply src/file_io.rs]: I ended up resorting to bit manipulation because it ends up being really gross and verbose with booleans. The logic you're representing ends up being weirdly tricky. An entry can be a file or a directory, hidden or not hidden, and can optionally be filtered by extension. And then the params allow you to specify that the entry can be hidden, not hidden, or both, etc. This ended up being the simplest way:
- A file is 0b01, a dir is 0b10; the param you provide is either 0b01, 0b10, or 0b11
- Not hidden is 0b01, hidden is 0b10; param is same as above
- Extension checked separately

File: src/fs_io.rs
Comment (howhelan) [inline src/fs_io.rs:110]: Thoughts on adopting a standard method comment style?  Method description, argument descriptions, and return value description.

**Review by howhelan — APPROVED** [review]

File: src/file_io.rs
Comment (howhelan) [reply src/file_io.rs]: Yea makes total sense!

File: src/file_io.rs
Comment (howhelan) [reply src/file_io.rs]: Fair enough, but I do think it could use a bit of method-level documentation.

Comment (ajpauwels) [reply src/file_io.rs]: you're right, will add a method-level comment. in fact, the code in general needs some more commenting.

File: src/fs_io.rs
Comment (ajpauwels) [reply src/fs_io.rs:27]: `self` is the concrete object that the method is attached to. in this case, `Metadata` is an enum type where the `0`'th element of the enum is an `fs::Metadata` type. `self.0` accesses that internally stored `fs::Metadata` object and called `is_file()` on it.

File: src/fs_io.rs
Comment (ajpauwels) [reply src/fs_io.rs:67]: Yes, this is the trait I plan to use for all current and future access to the concrete `fs` module. So right now the only function we're using is `read_dir`, but as our needs grow, any other function that needs to be mocked out, read or write, can be added here.

File: src/fs_io.rs
Comment (ajpauwels) [reply src/fs_io.rs:110]: open to any suggestions. we can add a doc to describe codebase standards.

File: src/file_io.rs
Comment (ajpauwels) [reply src/file_io.rs]: I messed up and added comments to a commit on the other PR that had other stuff on it.

Instead of taking the time to copy comments over one at a time, let's just merge this and take the comments on the other.

File: src/fs_io.rs
Comment (ajpauwels) [reply src/fs_io.rs:110]: Just remembered, no need to add all of that to the comments. Rust's doc generator is excellent and will extract all the info it needs out of the function signature itself. The only comment we need to remember to add is a prose block comment above the function with all the additional information/explanation you think necessary.

**Comment by capt-haddock** [comment]
@ajpauwels: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr | 03232070ff9dd74eb93b84ef227cc49c4ca108d6 | [link](http://dashboard.dev.pauwelslabs.com/namespaces/jx/pipelineruns/pr-p7wz8) | `/test pr`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by capt-haddock** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *[howhelan](https://github.com/pauwels-labs/redact-client/pull/14#pullrequestreview-649792123 "Approved")*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **capt-haddock**
You can assign the PR to them by writing `/assign @capt-haddock` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=pauwels-labs%2Fredact-client).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/pauwels-labs/redact-client/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by capt-haddock** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *[ajpauwels](https://github.com/pauwels-labs/redact-client/pull/14#issuecomment-832585098 "Approved")*, *[howhelan](https://github.com/pauwels-labs/redact-client/pull/14#pullrequestreview-649792123 "Approved")*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **capt-haddock**
You can assign the PR to them by writing `/assign @capt-haddock` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=pauwels-labs%2Fredact-client).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/pauwels-labs/redact-client/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by capt-haddock** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *[ajpauwels](https://github.com/pauwels-labs/redact-client/pull/14#issuecomment-832586558 "Approved")*, *[howhelan](https://github.com/pauwels-labs/redact-client/pull/14#pullrequestreview-649792123 "Approved")*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **capt-haddock**
You can assign the PR to them by writing `/assign @capt-haddock` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=pauwels-labs%2Fredact-client).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/pauwels-labs/redact-client/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
